File: LeetCode/src/linkedlist/RotateList.java
package linkedlist;

public class RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }

        ListNode cur = head;
        int length = 1;
        while (cur.next != null) {
            cur = cur.next;
            length++;
        }

        cur.next = head;

        int steps = length - k % length;
        ListNode tail = head;

        for (int i = 0; i < steps - 1; i++) {
            tail = tail.next;
        }

        ListNode newHead = tail.next;
        tail.next = null;

        return newHead;
    }

    public static void main(String[] args) {
        RotateList solution = new RotateList();
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        head1.next.next.next = new ListNode(4);
        head1.next.next.next.next = new ListNode(5);
        ListNode result1 = solution.rotateRight(head1, 2);
        printList(result1); // Output: 4->5->1->2->3

        ListNode head2 = new ListNode(0);
        head2.next = new ListNode(1);
        head2.next.next = new ListNode(2);
        ListNode result2 = solution.rotateRight(head2, 4);
        printList(result2); // Output: 2->0->1
    }

    public static void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
        System.out.println();
    }
}

File: LeetCode_Python/linkedlist/RotateList.py
from typing import Optional
from ListNode import ListNode


class RotateList:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or k == 0:
            return head

        length = 1
        cur = head

        while cur.next:
            cur = cur.next
            length += 1

        cur.next = head

        steps = length - k % length
        tail = head

        for _ in range(1, steps):
            tail = tail.next

        new_head = tail.next
        tail.next = None

        return new_head


# Time Complexity: O(n)
# Space Complexity: O(1)


if __name__ == "__main__":
    solution = RotateList()

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    print("Expected Output: 4 -> 5 -> 1 -> 2 -> 3")
    print("Actual Output:", end=" ")
    result = solution.rotateRight(head, 2)
    while result:
        if result.next:
            print(result.val, end=" -> ")
        else:
            print(result.val)
        result = result.next
